tests/ndc/ndc.go

```diff
@@ -0,0 +1,2468 @@
+				}},
+		}
+
+		s.standByReplicationTasksChan <- replicationTask
+		// this is to test whether dedup works
+		s.standByReplicationTasksChan <- replicationTask
+	}
+}
+
+func (s *NDCFunctionalTestSuite) eventBatchesToVersionHistory(
+	versionHistory *historyspb.VersionHistory,
+	eventBatches []*historypb.History,
+) *historyspb.VersionHistory {
+
+	// TODO temporary code to generate version history
+	//  we should generate version as part of modeled based testing
+	if versionHistory == nil {
+		versionHistory = versionhistory.NewVersionHistory(nil, nil)
+	}
+	for _, batch := range eventBatches {
+		for _, event := range batch.Events {
+			err := versionhistory.AddOrUpdateVersionHistoryItem(versionHistory,
+				versionhistory.NewVersionHistoryItem(
+					event.GetEventId(),
+					event.GetVersion(),
+				))
+			s.NoError(err)
+		}
+	}
+
+	return versionHistory
+}
+
+func (s *NDCFunctionalTestSuite) verifyEventHistorySize(
+	workflowID string,
+	runID string,
+	historySize int64,
+) {
+	// get replicated history events from passive side
+	describeWorkflow, err := s.cluster.GetFrontendClient().DescribeWorkflowExecution(
+		s.newContext(),
+		&workflowservice.DescribeWorkflowExecutionRequest{
+			Namespace: s.namespace.String(),
+			Execution: &commonpb.WorkflowExecution{
+				WorkflowId: workflowID,
+				RunId:      runID,
+			},
+		},
+	)
+	s.NoError(err)
+	// NOTE: non current branch can contain force termination event
+	//  so calculation should be updated, for now only assert below
+	s.True(historySize <= describeWorkflow.WorkflowExecutionInfo.HistorySizeBytes)
+}
+
+func (s *NDCFunctionalTestSuite) verifyVersionHistory(
+	workflowID string,
+	runID string,
+	expectedVersionHistory *historyspb.VersionHistory,
+) {
+	// get replicated history events from passive side
+	resp, err := s.cluster.GetHistoryClient().GetMutableState(
+		s.newContext(),
+		&historyservice.GetMutableStateRequest{
+			NamespaceId: string(s.namespaceID),
+			Execution: &commonpb.WorkflowExecution{
+				WorkflowId: workflowID,
+				RunId:      runID,
+			},
+		},
+	)
+	s.NoError(err)
+	if s.IsForceTerminated(workflowID, runID) {
+		index := resp.VersionHistories.CurrentVersionHistoryIndex
+		currentVersionHistory := resp.VersionHistories.Histories[index]
+		currentVersionHistory.Items[len(currentVersionHistory.Items)-1].EventId -= 1
+		// force termination event is generated by service itself, need to exclude
+	}
+	for _, actualVersionHistory := range resp.VersionHistories.Histories {
+		actualVersionHistory.BranchToken = nil
+	}
+	for _, actualVersionHistory := range resp.VersionHistories.Histories {
+		actualVersionHistory.BranchToken = nil
+		lcaItem, err := versionhistory.FindLCAVersionHistoryItem(
+			expectedVersionHistory,
+			actualVersionHistory,
+		)
+		s.NoError(err)
+		lastItem, err := versionhistory.GetLastVersionHistoryItem(expectedVersionHistory)
+		s.NoError(err)
+		if versionhistory.IsEqualVersionHistoryItem(lastItem, lcaItem) {
+			return
```

Comment:
As soon as you find lcaItem which equals expeectedHistory last item  this function exits. Should not this function be called something like: `assertItemInLCAHistory`?